### PR TITLE
at(typings): partial Types

### PR DESCRIPTION
**Pull request by Tenpi**
**Please describe the changes this PR makes and why it should be merged:**
This is a quick fix for Issue #3486 (Typescript users are forced to account for Partial types, even if they aren't using them). I separated partial properties into their own types such as:
```ts
type PartialMessage = {
    id: string;
    partial: true;
    fetch(): Promise<Message>;
} & {
    [P in keyof Omit<Message, "id" | "partial">]: null;
}
```
Made the following properties non-nullable (if any are wrong you can revert them): `message.author`.
And now all events will emit as `Message | PartialMessage`, `Channel | PartialChannel`, etc. 

```ts
client.on("message", async (message) => { 
    //message is Message | PartialMessage implicitly
    console.log(message.author.id) //message.author is possibly null
    if (message.partial) message = await message.fetch();
    console.log(message.author.id) //no errors
});

client.on("message", (message: Message) => {
    console.log(message.author.id) //message is Message, no errors
});
```
In separate functions, you must type message as `Message | PartialMessage` if you are using partials.

Finally, I deleted the type `PartialType` since it was an exact duplicate of `PartialTypes` (?)

Improvements: Emit partial types only if they are set in ClientOptions. I tried doing this, but I gave up since it was too complicated.  

**Status**
- [X] Code changes have been tested against the Discord API, or there are no code changes
- [ ] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [X] This PR changes the library's interface (methods or parameters added)
  - [X] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by Monbrey** [comment]
>  This PR only includes non-code changes, like changes to documentation, README, etc.

Not sure that's correct...

**Comment by Tenpi** [comment]
> > This PR only includes non-code changes, like changes to documentation, README, etc.
> 
> Not sure that's correct...

Sorry I fixed it... I wasn't too sure if modifying the typings fell under "code-changes".

**Comment by bdistin** [comment]
typings are documentation, not code changes.

File: typings/index.d.ts [hunk 1]
```diff
@@ -935,12 +935,12 @@ declare module 'discord.js' {
 		public editedTimestamp: number | null;
 		public readonly edits: Message[];
 		public embeds: MessageEmbed[];
-		public readonly guild: Guild | null;
+		public readonly guild: Guild;
```
Comment (vladfrangu) [inline typings/index.d.ts]: This is wrong. A guild can still be null (DM messages)

File: typings/index.d.ts [hunk 2]
```diff
@@ -2461,6 +2461,38 @@ declare module 'discord.js' {
 		| 'GUILD_MEMBER'
 		| 'MESSAGE';
 
+	type PartialMessage = {
+		id: string;
+		partial: true;
+		fetch(): Promise<Message>;
+	} & {
+		[P in keyof Omit<Message, 'id' | 'partial' | 'fetch'>]: null;
```
Comment (vladfrangu) [inline typings/index.d.ts]: Can't this be
```ts
[K in keyof Omit<Message, 'id' | 'partial>]: Message[K] | null;
```
? Also, why omit the function?

File: typings/index.d.ts [hunk 3]
```diff
@@ -2461,6 +2461,38 @@ declare module 'discord.js' {
...
+	};
+
+	type PartialChannel = {
+		id: string;
+		partial: true;
+		fetch(): Promise<Channel>;
+	} & {
+		[P in keyof Omit<DMChannel, 'id' | 'partial' | 'fetch'>]: null;
```
Comment (vladfrangu) [inline typings/index.d.ts]: This is wrong.. it doesn't extend just a DMChannel

File: typings/index.d.ts [hunk 4]
```diff
@@ -2461,6 +2461,38 @@ declare module 'discord.js' {
...
+	};
+
+	type PartialGuildMember = {
+		id: string;
+		partial: true;
+		fetch(): Promise<GuildMember>;
+	} & {
+		[P in keyof Omit<GuildMember, 'id' | 'partial' | 'fetch'>]: null;
```
Comment (vladfrangu) [inline typings/index.d.ts]: Ditto to my review on Guild

File: typings/index.d.ts [hunk 5]
```diff
@@ -2461,6 +2461,38 @@ declare module 'discord.js' {
...
+	};
+
+	type PartialUser = {
+		id: string;
+		partial: true;
+		fetch(): Promise<User>;
+	} & {
+		[P in keyof Omit<User, 'id' | 'partial' | 'fetch'>]: null;
```
Comment (vladfrangu) [inline typings/index.d.ts]: Ditto

**Comment by Tenpi** [comment]
Thanks for the review, without the Omit the partial property would be `true & null`, which actually makes it false so I had to add it to fix that bug.

**Comment by vladfrangu** [comment]
No worries, but I was asking why you Omitted the fetch function :sweat_smile:

**Comment by Tenpi** [comment]
> No worries, but I was asking why you Omitted the fetch function 😅

Oh yeah you're right, that was unnecessary

**Comment by Tenpi** [comment]
Thanks for the review, since `Partial<T>` is already a built-in typescript utility type I called it `Partialize<T>` instead. Tell me if you want another name ¯\\_(ツ)_/¯

**Comment by iCrawl** [comment]
Is `Partial` some reserved/taken keyword or why did you choose to go with `Partialize`?
I'd rather stick with Partial in this case.

**Comment by Skillz4Killz** [comment]
@Tenpi You are amazing. This is an awesome PR! Thank you so much!

**Comment by Skillz4Killz** [comment]
@Tenpi Would this or could it be adjusted to also solve #3533 ?

File: typings/index.d.ts [hunk 6]
```diff
@@ -836,12 +836,12 @@ declare module 'discord.js' {
 		public readonly displayName: string;
 		public guild: Guild;
 		public readonly id: Snowflake;
-		public readonly joinedAt: Date | null;
-		public joinedTimestamp: number | null;
+		public readonly joinedAt: Date;
+		public joinedTimestamp: number;
```
Comment (Lewdcario) [inline typings/index.d.ts]: Should be marked as null, due to https://github.com/discordjs/discord.js/commit/13bfceb83cf771d1d56d28fd474d824f9a8884db.

Comment (Tenpi) [reply typings/index.d.ts]: Updated, thanks for the explanation

**Pull request merged**